packages/compiler-core/src/codegen.ts

```diff
@@ -208,9 +208,27 @@ export function generate(
   }
 
   // binding optimizations
-  const optimizeSources = options.bindingMetadata
-    ? `, $props, $setup, $data, $options`
-    : ``
+  let optimizeSources = ''
+  const bindingMetadata = options.bindingMetadata
+  if (bindingMetadata) {
+    const bindingMetadataRank: MetadataType[] = [
+      'props',
+      'setup',
+      'data',
+      'options'
+    ]
+    const bindingMetadataValues = [...new Set(Object.values(bindingMetadata))]
+    let metadataArgs: MetadataType[] = []
+    for (let i = bindingMetadataRank.length - 1; i >= 0; i--) {
+      if (bindingMetadataValues.includes(bindingMetadataRank[i])) {
+        metadataArgs = bindingMetadataRank.slice(0, i + 1)
```

Comment:
Firstly, I think bindingMetadataRank can be hoisted out of the function since it will be used very frequently.
```javascript
const bindingMetadataRank= ['props',
      'setup',
      'data',
      'options']
const metadataArgs = bindingMetadataRank.map(metadataName => `$${metadataName}`)
```
Then the for loop could be done more efficiently to not allocate extra memory and with less time complexity.
```javascript
for (let i = bindingMetadataRank.length - 1; i >= 0; i--) {
  if (bindingMetadataRank[i] in bindingMetadata) {
    optimizeSources = `, ${metadataArgs.slice(0, i + 1).join(', ')}`
    break
  }
}
```

Reply:
I think the second suggestion is good,but the first may be not right,if i move `bindingMetadataRank` out of the `generate` function, it will affect other functions in this file while they don't need `bindingMetadataRank`.